app/data_client.go

```diff
@@ -761,3 +848,374 @@ func (d *DataClient) RemoveBinaryDataFromDatasetByIDs(
 	})
 	return err
 }
+
+func uploadMetadataToProto(metadata UploadMetadata) *syncPb.UploadMetadata {
+	methodParams, err := protoutils.ConvertMapToProtoAny(metadata.MethodParameters)
+	if err != nil {
+		return nil
+	}
+	return &syncPb.UploadMetadata{
+		PartId:           metadata.PartID,
+		ComponentType:    metadata.ComponentType,
+		ComponentName:    metadata.ComponentName,
+		MethodName:       metadata.MethodName,
+		Type:             syncPb.DataType(metadata.Type),
+		FileName:         metadata.FileName,
+		MethodParameters: methodParams,
+		FileExtension:    metadata.FileExtension,
+		Tags:             metadata.Tags,
+	}
+}
+
+func annotationsToProto(annotations Annotations) *pb.Annotations {
+	var protoBboxes []*pb.BoundingBox
+	for _, bbox := range annotations.Bboxes {
+		protoBboxes = append(protoBboxes, &pb.BoundingBox{
+			Id:             bbox.ID,
+			Label:          bbox.Label,
+			XMinNormalized: bbox.XMinNormalized,
+			YMinNormalized: bbox.YMinNormalized,
+			XMaxNormalized: bbox.XMaxNormalized,
+			YMaxNormalized: bbox.YMaxNormalized,
+		})
+	}
+	return &pb.Annotations{
+		Bboxes: protoBboxes,
+	}
+}
+
+func sensorMetadataToProto(metadata SensorMetadata) *syncPb.SensorMetadata {
+	return &syncPb.SensorMetadata{
+		TimeRequested: timestamppb.New(metadata.TimeRequested),
+		TimeReceived:  timestamppb.New(metadata.TimeReceived),
+		MimeType:      syncPb.MimeType(metadata.MimeType),
+		Annotations:   annotationsToProto(metadata.Annotations),
+	}
+}
+
+func sensorDataToProto(sensorData SensorData) *syncPb.SensorData {
+	protoSensorData := &syncPb.SensorData{
+		Metadata: sensorMetadataToProto(sensorData.Metadata),
+	}
+	switch {
+	case len(sensorData.SDBinary) > 0:
+		protoSensorData.Data = &syncPb.SensorData_Binary{
+			Binary: sensorData.SDBinary,
+		}
+	case sensorData.SDStruct != nil:
+		pbStruct, err := structpb.NewStruct(sensorData.SDStruct)
+		if err != nil {
+			return nil
+		}
+		protoSensorData.Data = &syncPb.SensorData_Struct{
+			Struct: pbStruct,
+		}
+	default:
+		return nil
+	}
+	return protoSensorData
```

Comment:
(q) I'm a little surprised by this. It definitely could be correct, but I would've expected the `default` case to also return the `protoSensorData`, so that the metadata is available if nothing else. Is that not correct? Is the `default` case something that will happen with any frequency, or is that something that we think will never occur?